UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,627 @@
+        } else {
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "compatible", &TempLen);
+          if (!(TempLen > 0)) {
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiReservedMemoryType);
+            continue;
+          }
+
+          TempStr = (CHAR8 *)(PropertyPtr->Data);
+
+          if (AsciiStrnCmp (TempStr, "boot-code", AsciiStrLen ("boot-code")) == 0) {
+            DEBUG ((DEBUG_INFO, "  boot-code"));
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiBootServicesCode);
+          } else if (AsciiStrnCmp (TempStr, "boot-data", AsciiStrLen ("boot-data")) == 0) {
+            DEBUG ((DEBUG_INFO, "  boot-data"));
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiBootServicesData);
+          } else if (AsciiStrnCmp (TempStr, "runtime-code", AsciiStrLen ("runtime-code")) == 0) {
+            DEBUG ((DEBUG_INFO, "  runtime-code"));
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiRuntimeServicesCode);
+          } else if (AsciiStrnCmp (TempStr, "runtime-data", AsciiStrLen ("runtime-data")) == 0) {
+            DEBUG ((DEBUG_INFO, "  runtime-data"));
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiRuntimeServicesData);
+          } else if (AsciiStrnCmp (TempStr, "acpi", AsciiStrLen ("acpi")) == 0) {
+            DEBUG ((DEBUG_INFO, "  acpi, StartAddress:%x, NumberOfBytes:%x", StartAddress, NumberOfBytes));
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiBootServicesData);
+            PlatformAcpiTable = BuildGuidHob (&gUniversalPayloadAcpiTableGuid, sizeof (UNIVERSAL_PAYLOAD_ACPI_TABLE));
+            if (PlatformAcpiTable != NULL) {
+              DEBUG ((DEBUG_INFO, " build gUniversalPayloadAcpiTableGuid , NumberOfBytes:%x", NumberOfBytes));
+              PlatformAcpiTable->Rsdp            = (EFI_PHYSICAL_ADDRESS)(UINTN)StartAddress;
+              PlatformAcpiTable->Header.Revision = UNIVERSAL_PAYLOAD_ACPI_TABLE_REVISION;
+              PlatformAcpiTable->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_ACPI_TABLE);
+            }
+          } else if (AsciiStrnCmp (TempStr, "acpi-nvs", AsciiStrLen ("acpi-nvs")) == 0) {
+            DEBUG ((DEBUG_INFO, "  acpi-nvs"));
+            BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiACPIMemoryNVS);
+          } else if (AsciiStrnCmp (TempStr, "smbios", AsciiStrLen ("smbios")) == 0) {
+            DEBUG ((DEBUG_INFO, " build smbios, NumberOfBytes:%x", NumberOfBytes));
+            SmbiosTable = BuildGuidHob (&gUniversalPayloadSmbios3TableGuid, sizeof (UNIVERSAL_PAYLOAD_SMBIOS_TABLE) + sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT));
+            if (SmbiosTable != NULL) {
+              SmbiosTable->Header.Revision  = UNIVERSAL_PAYLOAD_SMBIOS_TABLE_REVISION;
+              SmbiosTable->Header.Length    = sizeof (UNIVERSAL_PAYLOAD_SMBIOS_TABLE);
+              SmbiosTable->SmBiosEntryPoint = (EFI_PHYSICAL_ADDRESS)(UINTN)(&StartAddress);

```

Comment:
&StartAddress does not sound correct. Need to remove & here?

Reply:
resolved